libusb/descriptor.c

```diff
@@ -30,54 +30,17 @@
  * for detected devices
  */
 
-#define READ_LE16(p) ((uint16_t)	\
-	(((uint16_t)((p)[1]) << 8) |	\
-	 ((uint16_t)((p)[0]))))
-
-#define READ_LE32(p) ((uint32_t)	\
-	(((uint32_t)((p)[3]) << 24) |	\
-	 ((uint32_t)((p)[2]) << 16) |	\
-	 ((uint32_t)((p)[1]) <<  8) |	\
-	 ((uint32_t)((p)[0]))))
-
-static void parse_descriptor(const void *source, const char *descriptor, void *dest)
+static inline uint16_t ReadLittleEndian16(const uint8_t p[2])
 {
-	const uint8_t *sp = source;
-	uint8_t *dp = dest;
-	char field_type;
-
-	while (*descriptor) {
-		field_type = *descriptor++;
-		switch (field_type) {
-		case 'b':	/* 8-bit byte */
-			*dp++ = *sp++;
-			break;
-		case 'w':	/* 16-bit word, convert from little endian to CPU */
-			dp += ((uintptr_t)dp & 1);	/* Align to 16-bit word boundary */
-
-			*((uint16_t *)dp) = READ_LE16(sp);
-			sp += 2;
-			dp += 2;
-			break;
-		case 'd':	/* 32-bit word, convert from little endian to CPU (4-byte align dst before write). */
-			dp += 4 - ((uintptr_t)dp & 3);	/* Align to 32-bit word boundary */
+	return (uint16_t)((p[1] << 8) | p[0]);
+}
 
-			*((uint32_t *)dp) = READ_LE32(sp);
-			sp += 4;
-			dp += 4;
-			break;
-		case 'i':	/* 32-bit word, convert from little endian to CPU (no dst alignment before write) */
-			*((uint32_t *)dp) = READ_LE32(sp);
-			sp += 4;
-			dp += 4;
-			break;
-		case 'u':	/* 16 byte UUID */
-			memcpy(dp, sp, 16);
-			sp += 16;
-			dp += 16;
-			break;
-		}
-	}
+static inline uint32_t ReadLittleEndian32(const uint8_t p[4])
+{
+	return (uint32_t)((uint32_t)(p[3] << 24) |
+					  (uint32_t)(p[2] << 16) |
+					  (uint32_t)(p[1] << 8) |
```

Comment:
Shouldn't the cast/promotion be done before the shifting?

Reply:
Cast before shifting is not needed because of implicit promotion: https://www.gnu.org/software/c-intro-and-ref/manual/html_node/Shift-Operations.html (as long as we don't care about 16-bit platforms!).
I wouldn't think casting before bitwise or'ing is needed either.

Reply:
Indeed. I have restored the casting to be as it was in the macro.

Reply:
Why can't we just write
`return (p[3] << 24) | (p[2] << 16) | (p[1] << 8) | p[0];`
?

Reply:
That would implicitly promote to *signed* int (IIUC), and it seems safer to keep everything unsigned because [shifting a 1 into the sign bit is technically UB](https://blog.regehr.org/archives/738). Besides, it was already like that, and presumably correct as it was.

Reply:
Thanks for the link, interesting. I knew shifting right can be tricky but that shifting left is "undefined behavior" was new. In any case, for GCC there is no ambiguity: "Note: according to the C standard, shifting of signed values isn’t guaranteed to work properly when the value shifted is negative, or becomes negative during the operation of shifting left. However, only pedants have a reason to be concerned about this; only computers with strange shift instructions could plausibly do this wrong. In GNU C, the operation always works as expected" (https://www.gnu.org/software/c-intro-and-ref/manual/html_node/Shift-Caveats.html)

Actually there would be no -Wconversion warning without the casts either. But I agree to leave them in for correctness. The final cast is not needed though, it is implicit in return statements.

Reply:
>The final cast is not needed though, it is implicit in return statements.

C does lots of things implicitly, but I often find it's better to be explicit.

On a system where `int` is 64 bit (of which there are probably none that libusb supports) you might get a warning about potential truncation by *implicitly* converting from 64 to 32 bit.

But I just want this merged, so will remove it if you insist...